File: problem_0005.py
#!/usr/bin/python
"""2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?"""

from common import prime_factors

def smallest_product_of_primes_under_n(n):
	common_factors = []
	for i in range(2,n+1):
		i_prime_factors = prime_factors(i)
		common_factors = contains_and_expand(i_prime_factors, common_factors)
	
	product = 1
	for i in common_factors:
		product *= i
	return product

def contains_and_expand(list_a, list_b):
	for i in list_a:
		if i in list_b:
			list_b.remove(i)
	return list_a + list_b

if __name__ == "__main__":
	print smallest_product_of_primes_under_n(20)

File: problem_0005.hs
{-# LANGUAGE DeriveDataTypeable, RecordWildCards #-}

-- 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
-- What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

import Text.Printf
import System.Console.CmdArgs
import Data.Time

data EulerArgs = 
    Limits { 
        upper :: Integer,
        lower :: Integer } 
    | Factors { 
        factors :: [Integer] }
    | Euler 
        deriving (Show, Data, Typeable)
    
limits = Limits{ upper = 10, lower = 1 }
factors_ = Factors{ factors = [1..10] }
euler = Euler{}

problem0005 :: (Integral a) => [a] -> a 
problem0005 factors = product commonFactors
    where
        commonFactors = foldl1 combineLists factorList
        factorList = map (getFactors 2) factors

getFactors :: (Integral a) => a -> a -> [a]
getFactors _ 1 = [1]
getFactors f n
    | n < f * f = [n]
    | 0 == mod n f = f : getFactors f (quot n f)
    | otherwise = getFactors (succ f) n

combineLists :: (Integral a) => [a] -> [a] -> [a]
combineLists [] ys = ys
combineLists xs [] = xs
combineLists xs@(x:xs')  ys@(y:ys')
    | x < y = x : combineLists xs' ys
    | x > y = y : combineLists xs ys'
    | otherwise = x : combineLists xs' ys'

exec :: EulerArgs -> Integer
exec Limits{..} = problem0005 [lower..upper]
exec Factors{..} = problem0005 factors
exec Euler = problem0005 [1..20]

main :: IO ()
main = do
    args <- cmdArgs $ modes [limits, factors_, euler]
    start <- getCurrentTime
    print $ exec args
    stop <- getCurrentTime
    print $ diffUTCTime stop start

